File: sdk/typescript/src/kameleoLocalApiClient.ts
import * as apis from "./apis";
import { ErrorCode } from "./models";
import { Configuration, type ConfigurationParameters, DefaultConfig, ResponseError } from "./runtime";

export class KameleoLocalApiClient {
    /** Common configuration for all APIs, set at class construction. */
    public readonly configuration: Configuration;

    // APIs
    public readonly fingerprint: apis.FingerprintApi;
    public readonly cookie: apis.CookieApi;
    public readonly folder: apis.FolderApi;
    public readonly general: apis.GeneralApi;
    public readonly profile: apis.ProfileApi;
    public readonly kernel: apis.KernelApi;

    /**
     * Initializes a new instance of the KameleoLocalApiClient class.
     * @param configuration Configuration for connecting to the Kameleo Engine, including the base url.
     */
    public constructor(configuration?: ConfigurationParameters) {
        this.configuration = configuration ? new Configuration(configuration) : new Configuration();

        this.fingerprint = new apis.FingerprintApi(this.configuration);
        this.cookie = new apis.CookieApi(this.configuration);
        this.folder = new apis.FolderApi(this.configuration);
        this.general = new apis.GeneralApi(this.configuration);
        this.profile = new apis.ProfileApi(this.configuration);
        this.kernel = new apis.KernelApi(this.configuration);
    }

    /**
     * Verifies that the Kameleo Engine is ready to accept connections.
     * Throws if the engine is not available, or the engine was started, but does not become ready within the timeout.
     */
    public async verifyEngineReady(timeoutMs = 30_000): Promise<void> {
        const deadline = Date.now() + timeoutMs;
        while (true) {
            try {
                await this.general.healthcheck();
                return;
            } catch (e) {
                if (e instanceof ResponseError && e.errorCode === ErrorCode.ServiceNotReady && Date.now() < deadline) {
                    await new Promise((resolve) => setTimeout(resolve, 1_000));
                    continue;
                }

                const error = new Error(
                    "ERROR: Could not connect to Kameleo. Make sure it is running, or download it at https://kameleo.io/downloads",
                );

                // we target ES6, but Error.cause is only available in ~ES2022
                (error as any).cause = e;
                throw error;
            }
        }
    }
}

File: sdk/csharp/src/Kameleo.LocalApiClient/KameleoLocalApiClient.cs
using Kameleo.LocalApiClient.Api;
using Kameleo.LocalApiClient.Client;
using System;
using System.Threading;
using System.Threading.Tasks;

namespace Kameleo.LocalApiClient
{
    /// <summary>
    /// Encapsualte access to all the provided APIs by the Kameleo Engine.
    /// </summary>
    public class KameleoLocalApiClient : IKameleoLocalApiClient
    {
        private readonly Configuration _configuration;

        /// <inheritdoc/>
        public IReadableConfiguration Configuration => _configuration;

        /// <summary>
        /// Initializes a new instance of the <see cref="KameleoLocalApiClient"/> class using the default parameters.
        /// </summary>
        public KameleoLocalApiClient() : this(new Uri("http://localhost:5050"))
        {
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="KameleoLocalApiClient"/> class providing the base url of the Kameleo Engine.
        /// </summary>
        /// <param name="baseUri">Base url of the Kameleo Engine</param>
        public KameleoLocalApiClient(Uri baseUri) : this(new Configuration { BasePath = baseUri.ToString() })
        {
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="KameleoLocalApiClient"/> class using <see cref="Kameleo.LocalApiClient.Client.Configuration"/> object.
        /// </summary>
        /// <param name="configuration"><see cref="Kameleo.LocalApiClient.Client.Configuration"/></param>
        public KameleoLocalApiClient(Configuration configuration)
        {
            _configuration = configuration;

            Fingerprint = new FingerprintApi(_configuration);
            Cookie = new CookieApi(_configuration);
            Folder = new FolderApi(_configuration);
            General = new GeneralApi(_configuration);
            Profile = new ProfileApi(_configuration);
            Kernel = new KernelApi(_configuration);
        }

        /// <inheritdoc/>
        public IFingerprintApiAsync Fingerprint { get; }

        /// <inheritdoc/>
        public ICookieApiAsync Cookie { get; }

        /// <inheritdoc/>
        public IFolderApiAsync Folder { get; }

        /// <inheritdoc/>
        public IGeneralApiAsync General { get; }

        /// <inheritdoc/>
        public IProfileApi Profile { get; }

        /// <inheritdoc/>
        public IKernelApi Kernel { get; }

        /// <summary>
        /// Verifies that the Kameleo Engine is ready to accept connections.
        /// Throws if the engine is not available, or the engine was started, but does not become ready within the timeout.
        /// </summary>
        /// <param name="timeout">How long to wait for the engine to become ready. Defaults to 30 seconds.</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        public async Task VerifyEngineReadyAsync(TimeSpan? timeout = null, CancellationToken cancellationToken = default)
        {
            var deadline = DateTime.UtcNow + (timeout ?? TimeSpan.FromSeconds(30));
            while (true)
            {
                try
                {
                    await General.HealthcheckAsync(cancellationToken: cancellationToken).ConfigureAwait(false);
                    return;
                }
                catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
                {
                    throw;
                }
                catch (ApiException ex) when (ex.ErrorCode == Model.ErrorCode.ServiceNotReady && DateTime.UtcNow < deadline)
                {
                    await Task.Delay(1_000, cancellationToken).ConfigureAwait(false);
                }
                catch (Exception ex)
                {
                    throw new InvalidOperationException(
                        "ERROR: Could not connect to Kameleo. Make sure it is running, or download it at https://kameleo.io/downloads",
                        ex);
                }
            }
        }
    }
}
